### PR TITLE
464 - Add Datagrid Frozen Columns

### LINKED ISSUE #464: SOHO-1564: Datagrid Frozen Columns
The inforDataGrid allows you to specify a "frozenColumn" option, which causes some columns to become fixed so that when the grid is too wide to fit on the page, the scroll bar that appears only applies to the unfrozen columns.  Under the frozen columns, there is a scrollbar, but it is always disabled.
  It would be nice if there was some way to reduce the width of the region containing the frozen columns, and then using the scroll bar under the frozen columns to see them.

Landmark: Needs ability to freeze/lock the at least first column of a datagrid, which may be a possible workaround.

Sales Hub: Needs to freeze/lock the last column (which is an action button) that should always stay in view.

### LINKED ISSUE #1191: Datagrid Editable Component: Calendar for Order Date as alignment issues
**Describe the bug**
Datagrid Editable Component: Calendar for Order Date as alignment issues

**To Reproduce**
1. Go to url: http://master-enterprise.demo.design.infor.com/components/datagrid/example-editable
2. Click 3/15/2016 from the "Order Date" column
3. The calendar icon is appearing to close to the date
4. Please see screen shot attached. 

**Expected behavior**
Spacing is needed between the date and the calendar icon

**Screenshots**
datagrid calendar alignment issue

**Platform**
 - OS Version: [iOS 8]

**Additional context**
None

### LINKED ISSUE #1246: Datagrid - Display text for a button is not fully rendered
**Describe the bug**
Have a datagrid column that contains a button through the use of the `Formatters.Button`. When a `grid.hideColumn()` is performed, the display text on the button is not fully rendered. The button column is the last column in the grid.

**To Reproduce**
1. See screenshot below

**Expected behavior**
The button should be fully rendered.

**Version**
- ids-enterprise: 4.x.0

**Screenshots**
Landmark example:
image

**Platform**
 - Desktop browsers

**Additional context**
n/a

### LINKED ISSUE #1411: Datagrid column width error when filtering after resize
**Describe the bug**
Resizing a column then immediately applying a column level filter results in a formatting error

**To Reproduce**
1. Go to https://4140-beta0-enterprise.demo.design.infor.com/components/datagrid/test-tree-filter.html
3. Increase the size of the Task column.
3. Click on the Task column filter
4. Type 'hmm'
5. Press Enter.

**Expected behavior**
Column formatting should remain resized.

**Version**
- ids-enterprise: 4.14.0-beta.0

**Screenshots**
resizefilter

**Platform**
 - Windows 10
 - Chrome
 - Version 71.0.3578.98 (Official Build) (64-bit)

### LINKED ISSUE #1456: Datagrid: Smaller page sizes can not seem to scroll horizontally
**Describe the bug**
At smaller page sizes can not seem to scroll horizontally to view the last column with datagrid.

**To Reproduce**
1. Go to 'http://localhost:4000/components/datagrid/example-index.html'
2. Open page with full screen width
3. Resize the page so you can hide last column only
4. Try to scroll to reach that hidden last column
5. No scrolling happen

**Expected behavior**
Should let it scroll to see all the columns.

### LINKED ISSUE #1481: When using minWidth option the sort will misalign columns
**Describe the bug**
When using minWidth, changing the column width before reordering will cause the columns to misalign

**To Reproduce**
1. Go to https://codepen.io/anon/pen/REqWqz
2. Resize the "Product Name" column
3. Click "Product Name" Header (sort)
4. See Misaligned  columns

**Expected behavior**
Should Remain aligned

**Version**
- latest-enterprise --- 4.14.0

### LINKED ISSUE #1560: DataGrid: Selected Filter is not working as expected
**Describe the bug**
DataGrid: Selected Filter is not working as expected

**To Reproduce**
1. Go to >>
https://4150-beta0-enterprise.demo.design.infor.com/components/datagrid/example-filter-rowstatus.html 
OR 
https://4150-beta0-enterprise.demo.design.infor.com/components/datagrid/example-filter.html
2. In the table, notice that the UNCHECKED or NOT SELECTED is 4 rows
3. Filter 'In Stock' to NOT SELECTED
4. Only 2 columns show up

**Expected behavior**
Expecting 4 rows to be displayed. 

**Version**
 [4.150]

**Screenshots**
screen shot 2019-01-29 at 5 25 38 pm
screen shot 2019-01-30 at 11 39 07 am

**Platform**
- All Browser OS and Devices. 

**Additional context**
No additional context.

### PR DESCRIPTION
**Pull request by tmcconechy**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**

This PR adds a really cool frozen column feature to the datagrid. This has an impact on columns/column rendering since new containers are added. While fixing this i fixed other column rendering issues as listed. Also cleaned up the test pages a lot removing old ones or dups or one that did not make sense and added more tests for datagrid example pages.

Main impact of frozen columns is things like

- column rendering
- column widths
- column resize
- sort
- export
- plus integration with various other features

**Frozen Columns Works with**
* Alternate row shading
* Drilldown (can be frozen on the left)
* Export
* Selection (can freeze on the left)
* Row Status
* Data Grouping and Totals
* Empty Message (will stay centered)
* Summary Row (you can freeze columns on the left)
* Editing
* Expandable Row -> But the expandable row is only in the middle section
* Filtering - Filtered columns can be frozen
* Row Activation

**Frozen Columns Limitations**
* Wont work with columns with colspans
* Wont work with dynamically height rows (example-comments.html)
* You can’t resize frozen columns
* Frozen columns cant be moved or hidden from the personalize dialog
* Ids are required for the frozen columns and should be used in general to avoid bugs
* If freezing the group column all the group by columns will be frozen
* If the total sum of the frozen columns is wider than the device - then it's not possible to scroll. Will want a future solution
* Percent Columns - Percent column widths will not work on frozen sections at this time.
* Will not work with expandOnActivate setting
* Grouped Headers will not work with frozen columns at this time
* Row Reorder will not work with frozen columns at this time
* Will not work with nested grids
* Will not work with clickToSelect setting
* Will not yet work with tree grid

NOTE: If changing the mac setting "show scrollbars: always vs when scrolling". You need to reload the page to rerender properly.

**Related github/jira issue (required)**:
Closes #464 
Closes #1246
Closes #1560
Closes #1411
Closes #1191
Closes #1481
Closes #1456

**Steps necessary to review your pull request (required)**:

**Reviewing #464** 
http://localhost:4000/components/datagrid/example-frozen-columns
- scrolll up and down and side to side in the middle section
- try things like arrowing around, sorting, paging, export, personalize column (anything in the page)
- you may need to resize the browser less wide to see scrolling

Other pages showing some integration with other features
http://localhost:4000/components/datagrid/example-editable
http://localhost:4000/components/datagrid/example-empty-message
http://localhost:4000/components/datagrid/example-filter
http://localhost:4000/components/datagrid/example-frozen-columns
http://localhost:4000/components/datagrid/example-keyword-search
http://localhost:4000/components/datagrid/example-list
http://localhost:4000/components/datagrid/example-mixed-selection
http://localhost:4000/components/datagrid/example-summary-row

**Reviewing #1246**
http://localhost:4000/components/datagrid/test-button-not-fully-rendered.html
- scroll to right
- button should show in full

**Reviewing #1560**
http://localhost:4000/components/datagrid/example-filter.html
- toggle the three states on the filter on the in-stock column and make sure the filtered results are correct

**Reviewing #1411**
http://localhost:4000/components/datagrid/test-tree-filter.html
- follow this steps https://github.com/infor-design/enterprise/issues/1411#issuecomment-459448579

**Reviewing #1191**
http://localhost:4000/components/datagrid/example-editable
- edit the order date column and there should be enough room

**Reviewing  #1481**
http://localhost:4000/components/datagrid/example-index
- resize the product name shorter and then sort (alignment should stay)

**Reviewing   #1456**
http://localhost:4000/components/datagrid/example-index.html
- open page with full screen width
- Resize the page so you can hide last column only
- Try to scroll to reach that hidden last column -> should now be able to

### CODE REVIEW

**Comment by EdwardCoyle** [comment]
Replicating what @davidcarlsonberg is seeing:

feb-14-2019-5_08-pm

That might be tough to fix, depending on the containment though.

**Comment by tmcconechy** [comment]
I don’t get it yet. Can you screen shot the alignment issue. The video seems ok to me.

**Comment by davidcarlsonberg** [comment]
alignment_issue

**Comment by tmcconechy** [comment]
Ok I see, yes I had this on windows so this happens there two since it’s always visible.

Hmm can I css off the fact that scroll are are visible on Mac? Will have to check tomorrow

**Comment by tmcconechy** [comment]
I pushed an updated for the scrolling issues. Need a few more reviewers again.

File: src/components/datagrid/_datagrid.scss [hunk 1]
```diff
@@ -78,7 +78,7 @@ $datagrid-short-row-height: 23px;
   }
 
   .datagrid-body {
-    height: calc(100% - 40px);
+    height: calc(100% - 39px);
```
Comment (pwpatton) [inline src/components/datagrid/_datagrid.scss]: - 39 due to a border between frozen and unfrozen sections?

Comment (tmcconechy) [reply src/components/datagrid/_datagrid.scss]: It removes a border on the bottom on IE

**Comment by tmcconechy** [comment]
@davidcarlsonberg - yeah thats a known limitation. for performance i only size the columns on load not on page resize so it just has the original size. Been considering fixing that but didnt want to add page resize events. So its a separate consideration

**Pull request merged**